[Data views]: Add quick actions

ntsekouras opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Resolves: https://github.com/WordPress/gutenberg/issues/55096

This PR adds quick actions in data views items. The idea is that an item can have primary and secondary actions. The API was updated with a new `isEligible` property, which handles the rendering or not of an action per item. This is needed because not all of the items can have the same actions.


### Design notes
This PR will need some design tweaks - at least:
1. Handle how we want to show the actions if they have only primary or secondary actions. Do we want to reserve a space even if the item has no secondary actions? 🤔 
2. Check design in `grid` view






## Testing Instructions
1. Go to pages list in site editor and look around
4. Everything should be functional


## Screenshots or screencast <!-- if applicable -->
<img width="702" alt="Screenshot 2023-10-25 at 12 17 44 PM" src="https://github.com/WordPress/gutenberg/assets/16275880/4a03af2b-a84b-4673-a071-a4f1409eada6">

ntsekouras committed:
[Data views]: Add quick actions

oandregal commented on packages/edit-site/src/components/actions/trash-post.js:
```diff
@@ -17,6 +23,11 @@ export default function useMoveToTrashAction() {
 			label: __( 'Move to Trash' ),
+			type: ACTION_TYPES.primary,
+			icon: trash,
+			isEligible( { status } ) {
```
This made me think of this interaction in wp-admin:

https://github.com/WordPress/gutenberg/assets/583546/7ee9a7a3-ea07-4fa6-a30f-d14815ea7485

Perhaps we should add `deletePermanently` action and make it available for every page (just as secondary)?

`deletePermanently` would probably require a different interaction (modal to confirm). It could be valuable to showcase what actions can do.

ntsekouras replied:
Exactly! I didn't add any of those for now, because we can start adding actions incrementally, after we land the main PR. What you mentioned though and the `restore` should be a quick follow up.

oandregal commented on packages/edit-site/src/components/actions/view-actions.js:
```diff
@@ -0,0 +1,46 @@
+ */
+import { ACTION_TYPES } from './constants';
+
+export const viewPostAction = {
```
Nit pick. My brain favors homogeneity, so I think it'd be happier if we did either 1) define all actions in one file or 2) define one action per file.

ntsekouras replied:
I can do that. I'll update and move them to single file. If we need to, we can reconsider/split in the future.

oandregal commented on packages/edit-site/src/components/actions/view-actions.js:
```diff
@@ -0,0 +1,46 @@
+	label: __( 'View' ),
+	type: ACTION_TYPES.primary,
+	icon: external,
+	isEligible( post ) {
```
👍 on `isEligible`

oandregal commented on packages/edit-site/src/components/actions/view-actions.js:
```diff
@@ -0,0 +1,46 @@
+export const postRevisionsAction = {
+	id: 'view-post-revisions',
+	label: __( 'View revisions' ),
+	type: ACTION_TYPES.secondary,
```
What are your thoughts on `isQuickAction: boolean` instead of `type`? Or other way to put it: do you anticipate having tertiary/quaternary/etc. actions?

It sounds like `type` should be fine – I'm mostly curious.

ntsekouras replied:
Boolean sounds good for now. If we need more types in the future we can always change again. I'll update tomorrow.

oandregal commented on packages/edit-site/src/components/actions/view-actions.js:
```diff
@@ -0,0 +1,46 @@
+	type: ACTION_TYPES.primary,
+	icon: external,
+	isEligible( post ) {
+		return post.status !== 'trash';
```
I checked how this is currently managed in wp-admin and only `trash` posts don't have a preview (private or scheduled ones do, for example). It sounds good as it is.

oandregal commented on packages/edit-site/src/components/dataviews/view-list.js:
```diff
@@ -34,7 +34,7 @@ import { useMemo, Children, Fragment } from '@wordpress/element';
  */
 import { unlock } from '../../lock-unlock';
-import FieldActions from './field-actions';
+import ItemActions from './item-actions';
```
Ah, nice one: these are not _field_ actions but _record_ actions. I guess _item_ actions works as fine, and we are already using item in `getValue` and `render` as well. 👍

oandregal replied:
Actually, I wouldn't mind if we called this component just `<Actions />` and `item-actions.js` was moved to `actions/index.js`. Though, this is a private component not meant to be public, so I don't have a strong opinion.

ntsekouras replied:
We also have the `view-actions` so we could leave as is for now. I think the folder Riad had created was about maybe reusing these actions elsewhere? 🤔 We can reconsider in the future.

oandregal commented on packages/edit-site/src/components/dataviews/item-actions.js:
```diff
@@ -0,0 +1,74 @@
+								<MenuItem
+									key={ action.id }
+									onClick={ () => action.perform( item ) }
+									isDestructive={ action.isDesctructive }
```
```suggestion
									isDestructive={ action.isDestructive }
```

I haven't seen any action implementing this, so I presume it hasn't been used yet?

ntsekouras replied:
The trash post action does.

oandregal commented on packages/edit-site/src/components/dataviews/item-actions.js:
```diff
@@ -0,0 +1,74 @@
+						key={ action.id }
+						icon={ action.icon }
+						onClick={ () => action.perform( item ) }
+						isDestructive={ action.isDesctructive }
```
```suggestion
						isDestructive={ action.isDestructive }
```

ntsekouras replied:
Good catch here! I'll update manually because there is one more occurrence in code(not in this change set).

oandregal commented on packages/edit-site/src/components/dataviews/item-actions.js:
```diff
@@ -0,0 +1,74 @@
+
+export default function ItemActions( { item, actions } ) {
+	const { primaryActions, secondaryActions } = useMemo( () => {
+		return actions.reduce(
```
Personal preference: I'd favor using `actions.forEach` for this kind of operation.

oandregal reviewed (commented):
I presume you are waiting for design feedback to land this, I'll let others do that. From a code point of view, this could be merged – though I left some comments.

If you'd rather land parts of this PR before the design conversation settles, one thing we could do is merge everything but the `type` feature (more actions available, isEligible, renaming things).

ntsekouras committed:
address feedback part 1

ntsekouras committed:
move all action to single file

jameskoster commented:
> Handle how we want to show the actions if they have only primary or secondary actions

Good question, and I'd welcome feedback on it. That said we may not need to prioritise it here since all pages should have the 'Edit' and 'Rename' actions. 

Some minor notes:

* Trash icon doesn't need to be red.
* Would it be trivial to add an 'Undo' affordance to the trash confirmation snackbar?
* Buttons should be the 32px size variant.
* Should the "View" action open in a new tab as the icon suggests? It doesn't in wp-admin, so perhaps we should be consistent with that. If so we may need to consider a different icon.

ntsekouras committed:
address feedback

ntsekouras commented:
@jameskoster I addressed the feedback, but didn't handle this one:
>Would it be trivial to add an 'Undo' affordance to the trash confirmation snackbar?

Let's do this in a follow up, to get the actions API in. This way we can iterate and in parallel others could add more actions.

>Should the "View" action open in a new tab as the icon suggests?

The page opens in the same page for consistency with the current pages list. If you have an icon to suggest, we can update easily.

youknowriad reviewed (approved):
LGTM 👍
